File: packages/simple_intent_receiver/lib/src/intent_receiver.dart
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:simple_intent_receiver/src/pigeons/intent.g.dart';

class IntentReceiver extends IntentEvents {
  final _controller = StreamController<Intent>();
  int? _lastAdded;

  Stream<Intent> get events => _controller.stream;

  @override
  void onIntentReceived(int timestamp, Intent intent) {
    if (_lastAdded == null || timestamp > _lastAdded!) {
      _controller.add(intent);
    }
  }

  IntentReceiver.setUp({
    BinaryMessenger? binaryMessenger,
    String messageChannelSuffix = '',
  }) {
    IntentEvents.setUp(
      this,
      binaryMessenger: binaryMessenger,
      messageChannelSuffix: messageChannelSuffix,
    );
  }

  Future<void> dispose() async {
    await _controller.close();
  }
}

File: packages/simple_intent_receiver/android/src/main/kotlin/eu/weblibre/simple_intent_receiver/IntentReceiver.kt
package eu.weblibre.simple_intent_receiver

import io.flutter.plugin.common.BinaryMessenger
import eu.weblibre.simple_intent_receiver.pigeons.IntentEvents
import eu.weblibre.simple_intent_receiver.pigeons.Intent as PigeonIntent

class IntentReceiver(messenger: BinaryMessenger) {
    private val intentEvents: IntentEvents = IntentEvents(messenger)

    fun sendIntent(timestamp: Long, intent: PigeonIntent) {
        intentEvents.onIntentReceived(timestamp, intent) { }
    }
}
